src/mumble/pluginAPI/PluginComponents.h

```diff
@@ -0,0 +1,134 @@
+	error_t (*freeMemory)(void *pointer);
+
+
+	
+	// -------- Getter functions --------
+
+	/// Gets the connection ID of the server the user is currently active on (the user's audio output is directed at).
+	///
+	/// @param[out] connection A pointer to the memory location the ID should be written to
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then it is valid to access the
+	/// 	value of the provided pointer
+	error_t (*getActiveServerConnection)(MumbleConnection_t *connection);
+
+	/// Fills in the information about the local user.
+	///
+	/// @param connection The ID of the server-connection to use as a context
+	/// @param[out] userID A pointer to the memory the user's ID shall be written to
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the passed pointer
+	/// 	may be accessed
+	error_t (*getLocalUserID)(MumbleConnection_t connection, MumbleUserID_t *userID);
+
+	/// Fills in the information about the given user's name. The name allocated if this function returns STATUS_OK
+	/// has to be freed by a call to freeMemory eventually.
+	///
+	/// @param connection The ID of the server-connection to use as a context
+	/// @param userID The user's ID whose name should be obtained
+	/// @param[out] userName A pointer to where the pointer to the allocated string (C-ecoded) should be written to
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the passed pointer
+	/// 	may be accessed
+	error_t (*getUserName)(MumbleConnection_t connection, MumbleUserID_t userID, const char **userName);
+
+	/// Fills in the information about the given channel's name. The name allocated if this function returns STATUS_OK
+	/// has to be freed by a call to freeMemory eventually.
+	///
+	/// @param connection The ID of the server-connection to use as a context
+	/// @param channelID The channel's ID whose name should be obtained
+	/// @param[out] channelName A pointer to where the pointer to the allocated string (C-ecoded) should be written to
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the passed pointer
+	/// 	may be accessed
+	error_t (*getChannelName)(MumbleConnection_t connection, MumbleChannelID_t channelID, const char **channelName);
+
+	/// Gets an array of all users that are currently connected to the provided server. This array that is allocated if
+	/// this function returns STATUS_OK has to be freed by a call to freeMemory after it is no longer needed by the plugin.
+	///
+	/// @param connection The ID of the server-connection to use as a context
+	/// @param[out] users A pointer to where the pointer of the allocated array shall be written
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the passed pointer
+	/// 	may be accessed
+	error_t (*getAllUsers)(MumbleConnection_t connection, MumbleUserID_t **users);
```

Comment:
Is the end of the array indicated by a "null" userID? 
That's what teamspeak does, but we can only do that if 0 would be a value that can never appear.

Maybe we can return both.
Maybe

```
template<typename T>
struct MumbleArrayReturn_t {
size_t length;
T* data;
}
```
and

`error_t (*getAllUsers)(MumbleConnection_t connection, MumbleArrayReturn_t<MumbleUserID_t> *users);`
or
`error_t (*getAllUsers)(MumbleConnection_t connection, MumbleArrayReturn_t<MumbleUserID_t> **users);`

or make seperate typedefs for the different types of arrays that we will have.
Though that would not be plain C anymore, plugins written in C would have to do manual adjustments which is meh.
Or we could just define the structs manually without a template.

Reply:
> Is the end of the array indicated by a "null" userID?

Tbh I haven't thought about that yet.

We could also add another out-param for the size of the array, can't we?

Templates aren't `C` and can thus not be used for the API and creating a struct for each array type we might have leads us back to the problem of having an awful amount of structs we have to carry around.
Unless of course we were to create a single struct that will hold the size of the array and a `void*` to that array. I don't really want to use void-pointers though

Reply:
For every array a size-parameter should be in place now. Unless there are good reason to handle this differently, this will be the way to go :)